src/components/AuthForm.tsx

```diff
@@ -0,0 +1,269 @@
+import React, { useEffect, useState } from "react";
+import {
+  getAuth,
+  signInWithEmailAndPassword,
+  createUserWithEmailAndPassword,
+} from "firebase/auth";
+import { app } from "../firebase";
+import { collection, addDoc } from "firebase/firestore";
+import { db } from "../firebase";
+import { object, string } from "yup";
+import { useAuth } from "../contexts/AuthContext";
+import classes from "../styles/AuthForm.module.scss";
+
+interface AuthFormProps {
+  isSignIn: boolean;
+  setIsSignIn: (value: boolean) => void;
+}
+
+interface AuthData {
+  username?: string;
+  email: string;
+  password: string;
+  confirmPassword?: string;
+}
+
+type FormError = {
+  fieldName: keyof AuthData;
+  message: string;
+};
+
+const AuthForm: React.FC<AuthFormProps> = ({ isSignIn, setIsSignIn }) => {
+  const [authData, setAuthData] = useState<AuthData>({
+    email: "",
+    password: "",
+  });
+
+  const [errors, setErrors] = useState<FormError[]>([]);
+  const [error, setError] = useState<string>("");
+
+  const { login } = useAuth();
+
+  useEffect(() => {
+    setAuthData({
+      email: "",
+      password: "",
+      username: "",
+      confirmPassword: "",
+    });
+    setErrors([]);
+    setError("");
+  }, [isSignIn]);
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    setAuthData({ ...authData, [name]: value });
+  };
+
+  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    validateField(name, value);
+  };
+
+  const validateField = (name: string, value: string) => {
+    try {
+      const fieldSchema = object().shape({
+        username:
+          !isSignIn && name === "username"
```

Comment:
you had to go the TS route :D 
Lets make a string enum of Auth data fields, so we can reuse them here in the validation and everywhere in the app.